src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,571 @@
+    with open(file_path, "w") as file:
+        yaml.dump(data, file)
+
+
+def _fetch_assemblies_to_upload(
+        upload_env_key: str,
+        workspace_id: int,
+        collection_source_dir: str,
+        upload_file_ext: list[str],
+) -> Tuple[int, dict[str, str]]:
+    """
+    Help fetch assemblies to upload.
+    """
+    if upload_file_ext != UPLOAD_FILE_EXT:
+        raise ValueError(f"Currently only support assembly files to upload")
+
+    count = 0
+    wait_to_upload_assemblies = dict()
+
+    assembly_dirs = [
+        os.path.join(collection_source_dir, d)
+        for d in os.listdir(collection_source_dir)
+        if os.path.isdir(os.path.join(collection_source_dir, d))
+    ]
+
+    for assembly_dir in assembly_dirs:
+
+        assembly_file_list = [
+            f
+            for f in os.listdir(assembly_dir)
+            if os.path.isfile(os.path.join(assembly_dir, f))
+            and os.path.join(assembly_dir, f).endswith(tuple(upload_file_ext))
+        ]
+
+        if len(assembly_file_list) != 1:
+            raise ValueError(f"One and only one assembly file that ends with {upload_file_ext} "
+                    f"must be present in {assembly_dir} directory")
+
+        count += 1
+        assembly_name = assembly_file_list[0]
+
+        _, uploaded = _read_upload_status_yaml_file(upload_env_key, workspace_id, assembly_dir, assembly_name)
+
+        if uploaded:
+            print(
+                f"Assembly {assembly_name} already exists in workspace {workspace_id}. Skipping."
+            )
+            continue
+
+        wait_to_upload_assemblies[assembly_name] = assembly_dir
+
+    return count, wait_to_upload_assemblies
+
+
+def _prepare_skd_job_dir_to_upload(job_dir: str, wait_to_upload_assemblies: dict[str, str]) -> str:
+    """
+    Prepare SDK job directory to upload.
+    """
+    data_dir = _sanitize_data_dir(job_dir)
+    for assembly_file, assembly_dir in wait_to_upload_assemblies.items():
+        src_file = _get_source_file(assembly_dir, assembly_file)
+        dest_file = os.path.join(data_dir, assembly_file)
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
+
+    return data_dir
+
+
+def _create_entries_in_sourcedata_workspace(
+        upload_env_key: str,
+        workspace_id: int,
+        assembly_names: list[str],
+        assembly_name_to_dir: dict[str, str],
+        assembly_name_to_upa: dict[str, str],
+        output_dir: str,
+) -> list[str]:
+    """
+    Create a standard entry in sourcedata/workspace for each assembly.
+    Hardlink to the original assembly file in sourcedata to avoid duplicating the file.
+    Update the uploaded.yaml file in the genome directory with assembly names and upa info.
+    """
+    upas = list()
+    for assembly_name in assembly_names:
+        try:
+            assembly_dir = assembly_name_to_dir[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name}") from e
```

Comment:
I don't think this key error or the one below are worth catching. IIUC they can only occur if the programming is wrong

Reply:
Updated!

Reply:
:+1: